uaclient/clouds/identity.py

```diff
@@ -49,7 +49,33 @@ def get_cloud_type() -> Tuple[Optional[str], Optional[NoCloudTypeReason]]:
     return (None, NoCloudTypeReason.NO_CLOUD_DETECTED)
 
 
+class CloudFactoryError(Exception):
```

Comment:
Can't we inherit from `UserFacingError` and bake in the messages into each exception ?
In that way we would avoid the if checks on `_get_contract_token_from_cloud_identity` and we would only need to re-raise the exceptions there

Reply:
The purpose of this PR was to separate the exceptions thrown by the factory function from the UserFacingErrors. This allows us to interpret the Errors differently based on where we're calling the factory function. It also keeps the CLI-specific details (the messages) in the `cli` module, instead of buried in the `clouds.identity` module.

I don't mind the if checks on the type of CloudFactoryError because it makes it really obvious which situation causes which message.

I'm open to further discussion though :)

Reply:
That's a good point. No problem, we can keep the implementation as is